cpp/src/arrow/io/mman.h

```diff
@@ -93,13 +93,10 @@ static void* mmap(void* addr, size_t len, int prot, int flags, int fildes, off_t
   const DWORD protect = __map_mmap_prot_page(prot);
   const DWORD desiredAccess = __map_mmap_prot_file(prot);
 
-  const off_t maxSize = off + (off_t)len;
```

Comment:
MSVC defines `off_t` as
```
typedef long _off_t; // file offset value
```
`long` type width is 32 bits and it's not enough to keep 64 bits of size_t(unsigned long long).
File size over 4Gb needs more than 32 bits, that caused the issue.